### DISCUSSION TITLE
CPU Overload Issue with AzuraCast

### DISCUSSION THREAD

**Discussion by ckeurk** [AzuraCast/AzuraCast#6213] [Q&A]
Hello,

I am an AzuraCast user and I recently installed the platform on a new server with 8 CPU and 30.8 GB of RAM. I have noticed that after approximately a week of use, AzuraCast starts to become increasingly CPU intensive, eventually reaching a point where the CPU is overloaded and I receive a live error message.mp3. Currently, I have only one webradio in use on AzuraCast, and I use Stereo Tool as my audio processing software.

I am concerned about this CPU overload issue as it not only affects the quality of my webradio stream.
_it's a zen moment, but at times it's a rush_

https://user-images.githubusercontent.com/379099/232543524-4ab1e486-1f8c-4b7b-a64c-f3966de6e1b3.mp4

Can you provide me with advice on how to address this CPU overload issue with AzuraCast? Are there any configuration settings I can adjust to improve the platform's performance? Is there a line of code to force AzuraCast to resume normally? 

Style `docker-compose down`

Thank you in advance for your assistance.

**Comment by renatopinto89** [comment]
Hi @ckeurk 
I can immediately identify two major problems in this:
- You have a LOT of CPU Steal, really, a LOT. That means that the actual server where your VPS is hosted is oversold as all hell, making it's physical CPU struggling a lot to keep up. I would strongly recommend changing your VPS provider, which makes this happen:
- You're using StereoTool, which is known for being very CPU intensive.

As you can see, not only your VPS CPU cany keep up, having to process StereoTool on top of it is making things much, much worse.

You can always try restarting the VPS for a quick fix, but it will still build up that load eventually.

Kind regards

**Reply by robz0** [reply]
Hope it's ok to dump my thinking and findings here.

I came across this thread after experiencing horrible server lockups: - stream Stop/Start every few seconds; Can't access admin, can't ssh in to server - big pauses for many second and then it comes back for a few seconds. Eventually it gets so bad, you have to restart the server from cloud panel. 

I am thankful for this post #5575, which gave the 'top' (or htop?) command for seeing what processes are taking up resources on the virtual machine. 

In my case I found kswapd0, running high and working its way up to 100% for total lockout. 
restart..

I need to let it run a while as this kswapd0 builds up over time I believe. But it was definitely the problem in my case, as without it running, AzuraCast runs smoothly.

I followed a Digital Ocean guide to adjust swappiness and hopefully that will manage that kswap thing. Will have to wait a while to see result.  https://www.digitalocean.com/community/tutorials/how-to-add-swap-space-on-ubuntu-16-04#tweak-your-swap-settings
I think kswapd0 is about the virtual machine/swap memory and I guess nothing to do with Azuracast.

WHAT SPEC?
It leads to some interesting questions and observations for me though about what you need to run AzuraCast. 

I did try AzuraCast on Digital Ocean before but it started stuttering on 2vCPU and 2GB (I think or maybe 4gb? - $12/month version anyway). This was a lower spec than the recommended one-click on DO but that was too expensive for us as we are still testing and discovering this project.

I had explored AC even less then, so moved to IONOS which promised more ram/cpu/disk space for a lower price.  (setting up was not as easy - but got it working #6311)

But then, on Ionos, with double the spec, it started sticking again, which led to this deeper enquiry. 

An observation/note on @ckeurk spec, on this thread-  8 cores and 30GB should be more than enough to run AzuraCast i would have thought. Even with all the 'steal' indicated surely.  No?

My curent sytsem does not seem to go much beyond 60% at spikes, with mostly betwen 8 to 30%  - without that kswap thing - at the moment! Though, others may have other processes. However I guess kswapd0 will be common to other users, since I did not put it there and its about swap ram on the vm.

So the upshot is start from the 'top' command, (or htop on some installations) - as advised by @ItsMitch in #5575 and also by others in other posts about CPU/resource usage.

Since the container is running good at the moment on this current spec. I might have to re-test on DO smaller instance one time, now I know about the *top* command!

**Reply by robz0** [reply]
Although I do wonder if AzuraCast does have impact on kswap as there are some AutoDJ configs with affect ram/cpu/balance I know.

**Reply by renatopinto89** [reply]
@robz0 

It seems to me that you´re having a lot of bad luck with your Server hosting providers, both with DO and IONOS, CPU steal is something that should NEVER happen, and depending on the CPU that is installed in the actual node that´s powering the VPS, it´s very hit or miss.

Best way I can help you with in this case is torecommend the one I use with AzuraCast, affordable and no issues at all, Hetzner
I don´t know if I can put a refferal link in here but if GitHub allows me, you´ll have a €20 credit to try it out.

Here goes: https://hetzner.cloud/?ref=Pxqxiu8I41cs

Kind Regards

**Reply by robz0** [reply]
Thanks for the link @renatopinto89 . Hetzner price/spec looks really good. Not sure why I have not looked at them. Have tried everthing else!! Oracle, Digital Ocean and IONOS.   Hetzner certainly has competitive pricing and good bandwidth too- which is important for radio streaming to many listeners (we hope :) 
The link worked ok, tho I was not ready to jump - so hope it works when i am. 

Currently, since my post 12 hours ago, on this thread. All is good on IONOS. 
Not revisited DO. Am using that for S3 storage at the moment. 

I think I have been interested in:
portability of AzuraCast and it has proven itself brilliant for that.  I've managed to move it relatively  seamlessly at least once. 

Support(ability)
this community (you), the knowledge-base, docs and the devs are brilliant. (moving from Libretime!! no comparison)
also I think that going through the hassles with IONOS etc has given me insight into how to get it up and running happily and not feeling tied to any particular host. 

Relevant again to this thread, Scalability with s3, sftp, dropbox (have tested all now) is brilliant. 

I have already been recommending AzuraCast to our group - to use and support also.  I can do that even more confidently now. 
Great to know we can jump to Hetzner or the like also, in case it gets too much on IONOS or wherever. 
Thanks for the recommendation,  Looks good.

**Reply by robz0** [reply]
@renatopinto89 - I just saw your message again and wanted to add... 
My reference to steal was a  theoretical question about the OPs set-up and stats to which you responded. Thanks for your answer to that too.  Also, to be fair to DO and IONOS, I don't think this was an issue. I didn't check closely enough while it was on DO but on Ionos, AzuraCast reports none to neglible steal.

My issue was definitely kswapd0, as without it, the AzuraCast server as been ticking along nicely for 3 days now without locking up as before.

Thanks again
